src/or/main.c

```diff
@@ -1458,6 +1458,55 @@ initialize_periodic_events_cb(evutil_socket_t fd, short events, void *data)
   }
 }
 
+/* Helper function: Set up periodic_events[] that only matches the given
+ * roles. */
+static void
+setup_periodic_events_by_roles(uint32_t roles)
+{
+  for (int i = 0; periodic_events[i].name; ++i) {
+    periodic_event_item_t *item = &periodic_events[i];
+    if (item->roles & roles) {
+      periodic_event_setup(item);
+    }
+  }
+}
+
+/* Helper function: Set up all events in periodic_events[] and filter them by
+ * roles for which we are configured for. */
+static void
+setup_periodic_events(void)
+{
+  const or_options_t *options = get_options();
+
+  /* At this stage, we are booting up so we need to know for which roles this
+   * tor instance has. */
```

Comment:
Let's make just one bitfield here, and call setup_periodic_events_by_roles once?